command/job_restart.go

```diff
@@ -0,0 +1,955 @@
+				}
+				return nil
+			}
+		}(task))
+	}
+	return restarts.Wait().ErrorOrNil()
+}
+
+// stopAlloc stops an allocation and blocks until the replacement allocation is
+// running.
+func (c *JobRestartCommand) stopAlloc(alloc *api.Allocation) error {
+	shortAllocID := limit(alloc.ID, c.length)
+
+	c.ui.Output(fmt.Sprintf(
+		"    %s: Rescheduling allocation %q for group %q",
+		formatTime(time.Now()),
+		shortAllocID,
+		alloc.TaskGroup,
+	))
+
+	var q *api.QueryOptions
+	if c.noShutdownDelay {
+		q = &api.QueryOptions{
+			Params: map[string]string{"no_shutdown_delay": "true"},
+		}
+	}
+
+	// Stop allocation and wait for its replacement to be running or for a
+	// blocked evaluation that prevents placements for this task group to
+	// happen.
+	_, err := c.client.Allocations().Stop(alloc, q)
+	if err != nil {
+		return fmt.Errorf("Failed to stop allocation %q: %v", shortAllocID, err)
+	}
+
+	// errCh receives an error if anything goes wrong or nil when the
+	// replacement allocation is running.
+	// Use a buffered channel to prevent both goroutine from blocking trying to
+	// send a result back.
+	errCh := make(chan error, 1)
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	go c.monitorBlockedEvals(ctx, alloc, errCh)
+	go c.monitorReplacementAlloc(ctx, alloc, errCh)
+
+	// If we receive an error and nil it's safe to ignore the error since the
+	// nil result is what we are looking for.
+	return <-errCh
+}
+
+// monitorBlockedEvals searches for blocked evaluations for the allocation job.
+//
+// Returns an error in errCh if anything goes wrong or if there are placement
+// failures for the allocation task group.
+func (c *JobRestartCommand) monitorBlockedEvals(ctx context.Context, alloc *api.Allocation, errCh chan<- error) {
+	q := &api.QueryOptions{WaitIndex: 1}
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		default:
+		}
+
+		evals, qm, err := c.client.Jobs().Evaluations(alloc.JobID, q)
+		if err != nil {
+			errCh <- fmt.Errorf("Failed to retrieve evaluations for job %q: %v", alloc.JobID, err)
+			return
+		}
+
+		for _, eval := range evals {
+			select {
+			case <-ctx.Done():
+				return
+			default:
+			}
+
+			if eval.Status != api.EvalStatusBlocked {
+				continue
+			}
+
+			failures := eval.FailedTGAllocs[alloc.TaskGroup]
+			if failures != nil {
+				errCh <- fmt.Errorf("%s:\n%s",
+					jobRestartFailedToPlaceNewAllocation,
+					formatAllocMetrics(failures, false, strings.Repeat(" ", 4)),
+				)
+				return
+			}
+		}
+		q.WaitIndex = qm.LastIndex
+	}
+}
+
+// monitorReplacementAlloc waits for the allocation to have a follow-up
+// placement and for the new allocation be running.
+//
+// Returns an error in errCh if anything goes wrong or nil when the new
+// allocation is running.
+func (c *JobRestartCommand) monitorReplacementAlloc(ctx context.Context, alloc *api.Allocation, errCh chan<- error) {
+	q := &api.QueryOptions{WaitIndex: 1}
```

Comment:
Similar to the `monitorBlockedEvals`, I'd use the index we got back from stopping the alloc as the starting point. That ensures we're only getting new information here.

Reply:
(Technically doesn't _ensure_ new information because it could just timeout, but it helps!)

Reply:
Yeah, I don't think setting the index here will give us only _new information_, it just controls _when_ the request unblocks. More details in https://github.com/hashicorp/nomad/pull/16278#discussion_r1125169155.

Since the allocation stop index is guaranteed to happen before `monitorReplacementAlloc` I'm not sure if there's a difference between using `1` or the stop index 🤔